pkg/operator/controller/ingress_status.go

```diff
@@ -0,0 +1,102 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// syncIngressControllerStatus computes the current status of ic and therefrom
+// updates the syncIngressControllerStatus resource.
+func (r *reconciler) syncIngressControllerStatus(deployment *appsv1.Deployment, ic *operatorv1.IngressController) error {
+	selector, err := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return fmt.Errorf("deployment has invalid spec.selector: %v", err)
+	}
+
+	updated := ic.DeepCopy()
+	updated.Status.AvailableReplicas = deployment.Status.AvailableReplicas
+	updated.Status.Selector = selector.String()
+	updated.Status.Conditions = computeIngressStatusConditions(updated.Status.Conditions, deployment)
+	if !ingressStatusesEqual(updated.Status, ic.Status) {
+		if err := r.client.Status().Update(context.TODO(), updated); err != nil {
+			return fmt.Errorf("failed to update ingresscontroller status: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// computeIngressStatusConditions computes the ingress controller's current state.
+func computeIngressStatusConditions(conditions []operatorv1.OperatorCondition, deployment *appsv1.Deployment) []operatorv1.OperatorCondition {
+	availableCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.IngressControllerAvailableConditionType,
+		Status: operatorv1.ConditionUnknown,
+	}
+	if deployment.Status.AvailableReplicas > 0 {
+		availableCondition.Status = operatorv1.ConditionTrue
+	} else {
+		availableCondition.Status = operatorv1.ConditionFalse
+		availableCondition.Reason = "DeploymentUnavailable"
+		availableCondition.Message = fmt.Sprintf(
+			"%d Deployment replicas available, want %d",
+			deployment.Status.AvailableReplicas, deployment.Status.Replicas)
```

Comment:
This would be an appropriate message for the `Progressing` condition, but for `Available`, it is false simply because no deployment replicas are available, and the number wanted is irrelevant to the condition.  We can leave this as is though till we add the `Progressing` condition (this is similar to the situation with `computeOperatorStatusConditions`).

Reply:
@Miciah good point about the IngressController `Available` message.

> (this is similar to the situation with computeOperatorStatusConditions).

I believe a slight difference exists with the Operator `Available` condition. With the IngressController `Available` condition, if 0 Deployments exist then status is `False`. For Operator, each IngressController must have at least 1 Deployment or status is `False`.

Reply:
Right, I mean the situation is similar in that we are conflating `Available` and `Progressing`.  Here, we are only reporting `Available`, but we are reporting a message that is more appropriate for `Progressing` (which does not exist yet).  In the operator status, we are reporting both `Available` and `Progressing`, but we report the same status and message, which are both more appropriate for `Progressing`.

Once we add a `Progressing` condition to `IngressController`, we can sort out the ingresscontroller conditions here, and then it will be possible to sort out the operator conditions too.  I just want to make sure we're on the same page, and I think we are.